testutils/resourcebuilder/machine/v1beta1/aws_provider_spec.go

```diff
@@ -71,44 +117,25 @@ func (m AWSProviderSpecBuilder) Build() *machinev1beta1.AWSMachineProviderConfig
 			APIVersion: "awsproviderconfig.openshift.io/v1beta1",
 			Kind:       "AWSMachineProviderConfig",
 		},
-		AMI: machinev1beta1.AWSResourceReference{
-			ID: ptr.To[string]("aws-ami-12345678"),
-		},
-		BlockDevices: []machinev1beta1.BlockDeviceMappingSpec{
-			{
-				EBS: &machinev1beta1.EBSBlockDeviceSpec{
-					Encrypted:  ptr.To[bool](true),
-					VolumeSize: ptr.To[int64](120),
-					VolumeType: ptr.To[string]("gp3"),
-				},
-			},
-		},
-		CredentialsSecret: &corev1.LocalObjectReference{
-			Name: "aws-cloud-credentials",
-		},
-		IAMInstanceProfile: &machinev1beta1.AWSResourceReference{
-			ID: ptr.To[string]("aws-iam-instance-profile-12345678"),
-		},
-		InstanceType: m.instanceType,
-		LoadBalancers: []machinev1beta1.LoadBalancerReference{
-			{
-				Type: "network",
-				Name: "aws-nlb-int",
-			},
-			{
-				Type: "network",
-				Name: "aws-nlb-ext",
-			},
-		},
-		Placement: machinev1beta1.Placement{
-			Region:           "us-east-1",
-			AvailabilityZone: m.availabilityZone,
-		},
-		SecurityGroups: m.securityGroups,
-		Subnet:         m.subnet,
-		UserDataSecret: &corev1.LocalObjectReference{
-			Name: "aws-user-data-12345678",
-		},
+		AMI: CoalesceAWSResourceReference(m.ami, machinev1beta1.AWSResourceReference{
```

Comment:
The generic Coalesce you wrote doesn't work here?

Reply:
Unfortunately not. That generic Coalesce relies on the provided type being comparable by generics standards.
Slice, map, and function types are not comparable, and so structs that contains them.

so they yield errors like so:
```
[]"github.com/openshift/api/machine/v1beta1".BlockDeviceMappingSpec does not satisfy comparable compiler[InvalidTypeArg](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#InvalidTypeArg)
```

Thus I've added some custom Coalesce functions for the types that are not adhering to comparable.

---

Another, more flexible option is to go with `interface{}` rather than generics, WDYT?:

```
// Coalesce returns the first value (value) provided it is a non-zero value for the type,
// otherwise it returns a the second value (defaultValue).
func Coalesce(i interface{}, defaultValue interface{}) interface{} {
	// Use reflection to get the underlying type and value
	val := reflect.ValueOf(i)

	// Check if the interface{} is a pointer
	if val.Kind() != reflect.Ptr {
		panic("Coalesce only accepts pointers to a value")
	}

	// Check if the pointer is nil
	if val.IsNil() {
		return defaultValue
	}

	// Dereference the pointer and return the value it points to
	return val.Elem().Interface()
}
```

And then use it like so:
```
LoadBalancers:      resourcebuilder.Coalesce(m.loadBalancers, defaultLoadBalancers).([]machinev1beta1.LoadBalancerReference),
Placement: resourcebuilder.Coalesce(m.placement, machinev1beta1.Placement{
	Region:           resourcebuilder.Coalesce(m.region, defaultRegion).(string),
	AvailabilityZone: resourcebuilder.Coalesce(m.availabilityZone, defaultAvailabilityZone).(string),
}).(machinev1beta1.Placement),
```

Reply:
No strong opinion either way to be honest, happy to leave as is, but the internal coalesce should be private right?

Reply:
Ok, went with generics.